File: cpp/problems/leetcode/0152_maximum_product_subarray/solution.cpp
// LeetCode 152
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if (nums.empty()) return 0;
        int mx = nums[0], mn = nums[0], best = nums[0];
        for (size_t i = 1; i < nums.size(); ++i) {
            int x = nums[i];
            int a = mx * x, b = mn * x;
            mx = max({x, a, b});
            mn = min({x, a, b});
            best = max(best, mx);
        }
        return best;
    }
};

int main() {
    vector<int> a{2, 3, -2, 4};
    assert(Solution().maxProduct(a) == 6);
    vector<int> b{-2, 0, -1};
    assert(Solution().maxProduct(b) == 0);
    vector<int> c{-2, 3, -4};
    assert(Solution().maxProduct(c) == 24);
    cout << "0152 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0152_maximum_product_subarray/solution.py
"""LeetCode 152: Maximum Product Subarray."""

from __future__ import annotations


class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        if not nums:
            return 0
        mx = mn = best = nums[0]
        for x in nums[1:]:
            cand = (x, mx * x, mn * x)
            mx = max(cand)
            mn = min(cand)
            best = max(best, mx)
        return best


if __name__ == "__main__":
    assert Solution().maxProduct([2, 3, -2, 4]) == 6
    assert Solution().maxProduct([-2, 0, -1]) == 0
    assert Solution().maxProduct([-2, 3, -4]) == 24
    print("0152 OK")
